Expandable chat toolbar

javitrujillo opened the pull request:
- Now the max number of lines displayed in the input text view is five.
- The chat toolbar can now be expanded when there are more than five lines in the input text view.
- The upper gap between the input text view and the toolbar is bigger.

Issue: https://issues.developers.mega.co.nz/issues/13749
InVision project: https://projects.invisionapp.com/d/main/#/console/18183157/377446623/preview#project_console
Freehand: https://projects.invisionapp.com/freehand/document/fT3cJzQzN

javitrujillo committed:
New UI settings for the input toolbar

- Maximum of 5 lines instead of 3.
- 6 additional pixels of separation between the top of the input toolbar and the top of its container.

javitrujillo committed:
Align the send button to the bottom instead to the top

- Show the send button instead of the microphone as soon as there is any text, including whitespaces (disabled until there is text that can be sent).

javitrujillo committed:
Added button to expand or collapse the chat toolbar

- Added two modes to the toolbar to track if it is collapsed or expanded.
- If pressing the expand button when the keyboard is not present, present it without expanding the toolbar.
- Show only the expand button when there are, at least, 5 lines of text in the input text view.
- Removed legacy code used to avoid a bug with the send button constraints. As the design changed with voice clips, it is no longer needed.
- Removed unused outlet.

javitrujillo committed:
Added draggable area with pan gesture to collapse the toolbar, if expanded

- Modify the draggable area height instead of the previous introduced constraint when changing from the collapsed mode to the expanded one.
- Extracted and renamed some constants out of a method to have them available in the entire class.
- Added pan gesture recognizer.
- Animate the change of the height constraint of the toolbar.
- Removed unused constant.

cmacera committed:
Merge branch 'chatroom_new_design' into feature/expandable_chat_toolbar

# Conflicts:
#	JSQMessagesViewController/Views/MEGAInputToolbar.m

cmacera committed:
Round corners of input text area in expanded state.

cmacera reviewed (approved):
The functionality is ok but I have found a minor related issues and other ones non related (minor UI issues already in production) to this PR. Due the upcoming replacement of JSQMessagesViewController vendor, I think it does not worth spend time on fixing those things and we can merge it in order to have the expandable text area what is an important request.

cmacera committed:
Added UI shadow view when text input area is expanded.

javisanesp commented:
The issue @jcmartinac is talking about is that each time you tap on the button to expand or collapse the expandable area, the messages begin to go up on the view. I agree with him that due the circumstances, we can ignore it. I did found two issues that we have to fix.

|  If you try to change to another action | Glitch when entering or deleting text   | 
| ------------- |:-------------:| 
|  <img src="https://user-images.githubusercontent.com/5855504/73451684-7b267880-4368-11ea-8ffc-873c7585faa5.gif" width=320>   |    <img src="https://user-images.githubusercontent.com/5855504/73451695-85487700-4368-11ea-92f3-a0a5386af247.gif" width=320>   |

javisanesp commented:
We also found that the tap area of the expand or collapse button is small. I'm going to increase it to the same that for accessory buttons (30x30).

javisanesp committed:
Some fixes

- Collapse the input toolbar if it is expanded when you tap on an accessory button.
- Deleted animation in the method to resize the toolbar, it caused a glitch when entering on the chat room or when deleting text.
- Increased area of the expand or collapse button to 30x30.
